File: mobile/src/main/java/com/backdoor/moove/PlacesListActivity.java
package com.backdoor.moove;

import android.app.AlertDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.backdoor.moove.core.adapters.PlaceRecyclerAdapter;
import com.backdoor.moove.core.consts.Constants;
import com.backdoor.moove.core.consts.QuickReturnViewType;
import com.backdoor.moove.core.data.PlaceDataProvider;
import com.backdoor.moove.core.helper.Coloring;
import com.backdoor.moove.core.helper.DataBase;
import com.backdoor.moove.core.helper.Module;
import com.backdoor.moove.core.helper.Permissions;
import com.backdoor.moove.core.interfaces.SimpleListener;
import com.backdoor.moove.core.utils.LocationUtil;
import com.backdoor.moove.core.utils.QuickReturnUtils;
import com.backdoor.moove.core.views.ReturnScrollListener;

public class PlacesListActivity extends AppCompatActivity implements SimpleListener {

    private RecyclerView listView;
    private LinearLayout emptyItem;
    private Coloring cs = new Coloring(PlacesListActivity.this);
    private FloatingActionButton mFab;

    private PlaceDataProvider provider;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTheme(cs.getStyle());
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().setStatusBarColor(cs.colorPrimaryDark());
        }
        setContentView(R.layout.places_activity_layout);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp);
        toolbar.setTitle(getString(R.string.places));

        emptyItem = findViewById(R.id.emptyItem);
        emptyItem.setVisibility(View.VISIBLE);

        TextView emptyText = findViewById(R.id.emptyText);
        emptyText.setText(getString(R.string.no_places));

        ImageView emptyImage = findViewById(R.id.emptyImage);
        emptyImage.setImageResource(R.drawable.place_white);

        listView = findViewById(R.id.currentList);

        mFab = findViewById(R.id.fab);
        mFab.setOnClickListener(v -> {
            if (LocationUtil.playServicesFullCheck(PlacesListActivity.this)) {
                if (Permissions.checkPermission(PlacesListActivity.this, Permissions.ACCESS_COARSE_LOCATION)) {
                    startActivity(new Intent(PlacesListActivity.this, NewPlaceActivity.class));
                } else {
                    Permissions.requestPermission(PlacesListActivity.this, 101, Permissions.ACCESS_COARSE_LOCATION,
                            Permissions.ACCESS_FINE_LOCATION);
                }
            }
        });
    }

    private void loadPlaces() {
        provider = new PlaceDataProvider(this, true);
        reloadView();
        PlaceRecyclerAdapter adapter = new PlaceRecyclerAdapter(this, provider, false);
        adapter.setEventListener(this);
        listView.setLayoutManager(new LinearLayoutManager(this));
        listView.setAdapter(adapter);
        listView.setItemAnimator(new DefaultItemAnimator());
        ReturnScrollListener scrollListener = new
                ReturnScrollListener.Builder(QuickReturnViewType.FOOTER)
                .footer(mFab)
                .minFooterTranslation(QuickReturnUtils.dp2px(this, 88))
                .isSnappable(true)
                .build();

        if (Module.isLollipop()) {
            listView.addOnScrollListener(scrollListener);
        } else {
            listView.setOnScrollListener(scrollListener);
        }
    }

    private void reloadView() {
        int size = provider.getCount();
        if (size > 0) {
            listView.setVisibility(View.VISIBLE);
            emptyItem.setVisibility(View.GONE);
        } else {
            listView.setVisibility(View.GONE);
            emptyItem.setVisibility(View.VISIBLE);
        }
    }

    private void deletePlace(int position) {
        long id = provider.getItem(position).getId();
        if (id != 0) {
            DataBase db = new DataBase(this);
            db.open();
            db.deletePlace(id);
            db.close();
            Snackbar.make(mFab, R.string.deleted, Snackbar.LENGTH_LONG)
                    .show();
            loadPlaces();
        }
    }

    private void editPlace(int position) {
        startActivity(new Intent(this, NewPlaceActivity.class)
                .putExtra(Constants.ITEM_ID_INTENT, provider.getItem(position).getId()));
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        loadPlaces();
    }

    @Override
    public void onItemClicked(int position, View view) {
        editPlace(position);
    }

    @Override
    public void onItemLongClicked(final int position, View view) {
        final CharSequence[] items = {getString(R.string.edit), getString(R.string.delete)};
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setItems(items, (dialog, item) -> {
            dialog.dismiss();
            if (item == 0) {
                editPlace(position);
            }
            if (item == 1) {
                deletePlace(position);
            }
        });
        AlertDialog alert = builder.create();
        alert.show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (grantResults.length == 0) return;
        switch (requestCode) {
            case 101:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    startActivity(new Intent(PlacesListActivity.this, NewPlaceActivity.class));
                } else {
                    Permissions.showInfo(PlacesListActivity.this, Permissions.READ_CALENDAR);
                }
                break;
        }
    }
}

File: mobile/src/main/java/com/backdoor/moove/PlacesListActivity.kt
package com.backdoor.moove

import android.app.AlertDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.appcompat.widget.Toolbar
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import com.backdoor.moove.modern_ui.places.list.PlacesAdapter
import com.backdoor.moove.core.consts.Constants
import com.backdoor.moove.core.consts.QuickReturnViewType
import com.backdoor.moove.core.data.PlaceDataProvider
import com.backdoor.moove.utils.Coloring
import com.backdoor.moove.core.helper.DataBase
import com.backdoor.moove.core.helper.Module
import com.backdoor.moove.core.helper.Permissions
import com.backdoor.moove.core.interfaces.SimpleListener
import com.backdoor.moove.core.utils.LocationUtil
import com.backdoor.moove.core.utils.QuickReturnUtils
import com.backdoor.moove.core.views.ReturnScrollListener

class PlacesListActivity : AppCompatActivity(), SimpleListener {

    private var listView: RecyclerView? = null
    private var emptyItem: LinearLayout? = null
    private var mFab: FloatingActionButton? = null

    private var provider: PlaceDataProvider? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val cs = Coloring(this@PlacesListActivity)
        setTheme(cs.style)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.statusBarColor = cs.colorPrimaryDark()
        }
        setContentView(R.layout.places_activity_layout)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        if (supportActionBar != null) supportActionBar!!.setDisplayShowTitleEnabled(false)
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp)
        toolbar.title = getString(R.string.places)

        emptyItem = findViewById(R.id.emptyItem)
        emptyItem!!.visibility = View.VISIBLE

        val emptyText = findViewById<TextView>(R.id.emptyText)
        emptyText.text = getString(R.string.no_places)

        val emptyImage = findViewById<ImageView>(R.id.emptyImage)
        emptyImage.setImageResource(R.drawable.place_white)

        listView = findViewById(R.id.currentList)

        mFab = findViewById(R.id.fab)
        mFab!!.setOnClickListener { v ->
            if (LocationUtil.playServicesFullCheck(this@PlacesListActivity)) {
                if (Permissions.checkPermission(this@PlacesListActivity, Permissions.ACCESS_COARSE_LOCATION)) {
                    startActivity(Intent(this@PlacesListActivity, NewPlaceActivity::class.java))
                } else {
                    Permissions.requestPermission(this@PlacesListActivity, 101, Permissions.ACCESS_COARSE_LOCATION,
                            Permissions.ACCESS_FINE_LOCATION)
                }
            }
        }
    }

    private fun loadPlaces() {
        provider = PlaceDataProvider(this, true)
        reloadView()
        val adapter = PlacesAdapter(this, provider, false)
        adapter.eventListener = this
        listView!!.layoutManager = LinearLayoutManager(this)
        listView!!.adapter = adapter
        listView!!.itemAnimator = DefaultItemAnimator()
        val scrollListener = ReturnScrollListener.Builder(QuickReturnViewType.FOOTER)
                .footer(mFab)
                .minFooterTranslation(QuickReturnUtils.dp2px(this, 88))
                .isSnappable(true)
                .build()

        if (Module.isLollipop) {
            listView!!.addOnScrollListener(scrollListener)
        } else {
            listView!!.setOnScrollListener(scrollListener)
        }
    }

    private fun reloadView() {
        if (provider == null) return
        val size = provider!!.count
        if (size > 0) {
            listView!!.visibility = View.VISIBLE
            emptyItem!!.visibility = View.GONE
        } else {
            listView!!.visibility = View.GONE
            emptyItem!!.visibility = View.VISIBLE
        }
    }

    private fun deletePlace(position: Int) {
        if (provider == null) return
        val id = provider!!.getItem(position)!!.id
        if (id != 0L) {
            val db = DataBase(this)
            db.open()
            db.deletePlace(id)
            db.close()
            Snackbar.make(mFab!!, R.string.deleted, Snackbar.LENGTH_LONG)
                    .show()
            loadPlaces()
        }
    }

    private fun editPlace(position: Int) {
        if (provider == null) return
        startActivity(Intent(this, NewPlaceActivity::class.java)
                .putExtra(Constants.ITEM_ID_INTENT, provider!!.getItem(position)!!.id))
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onResume() {
        super.onResume()
        loadPlaces()
    }

    override fun onItemClicked(position: Int, view: View) {
        editPlace(position)
    }

    override fun onItemLongClicked(position: Int, view: View) {
        val items = arrayOf<CharSequence>(getString(R.string.edit), getString(R.string.delete))
        val builder = AlertDialog.Builder(this)
        builder.setItems(items) { dialog, item ->
            dialog.dismiss()
            if (item == 0) {
                editPlace(position)
            }
            if (item == 1) {
                deletePlace(position)
            }
        }
        val alert = builder.create()
        alert.show()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (grantResults.size == 0) return
        when (requestCode) {
            101 -> if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                startActivity(Intent(this@PlacesListActivity, NewPlaceActivity::class.java))
            } else {
                Permissions.showInfo(this@PlacesListActivity, Permissions.READ_CALENDAR)
            }
        }
    }
}
